### ISSUE TITLE
Using `virt-clone` with `virt-lightning`-created VM

### ISSUE THREAD

**Issue by perlun** [virt-lightning/virt-lightning#322]
Hi,

I am trying to set up a GitLab Runner configuration using their custom executor support, similar to this: https://docs.gitlab.com/runner/executors/custom_examples/libvirt.html

The general idea is that I'll provision a set of BSD (NetBSD, FreeBSD, OpenBSD) machines using `virt-lightning` + Ansible playbook/role, then once all the required software for my CI jobs is available shutdown the VM. For each CI job, the executor can then clone it like this to provide a fresh VM for each run:

```shell
$ sudo virt-clone -o netbsd-100-ufs -n temp --auto-clone
$ sudo virt-clone -o netbsd-100-ufs -n temp2 --auto-clone
$ sudo virsh start temp
$ sudo virsh start temp2
```

This seems to work fine, **but** each VM gets the same IP address. `vl console temp` and `vl console temp2` gives me this:

### `temp` ifconfig output

```
netbsd-100-ufs# ifconfig
vioif0: flags=0x8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
        ec_capabilities=0x1<VLAN_MTU>
        ec_enabled=0
        address: 52:54:00:84:9e:74
        status: active
        inet6 fe80::5054:ff:fe84:9e74%vioif0/64 flags 0 scopeid 0x1
        inet 192.168.123.5/24 broadcast 192.168.123.255 flags 0x2<DUPLICATED>
lo0: flags=0x8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 33624
        status: active
        inet6 ::1/128 flags 0x20<NODAD>
        inet6 fe80::1%lo0/64 flags 0 scopeid 0x2
        inet 127.0.0.1/8 flags 0
```

### `temp2` ifconfig output

```
netbsd-100-ufs# ifconfig
vioif0: flags=0x8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
        ec_capabilities=0x1<VLAN_MTU>
        ec_enabled=0
        address: 52:54:00:7a:b8:64
        status: active
        inet6 fe80::5054:ff:fe7a:b864%vioif0/64 flags 0 scopeid 0x1
        inet 192.168.123.5/24 broadcast 192.168.123.255 flags 0
lo0: flags=0x8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 33624
        status: active
        inet6 ::1/128 flags 0x20<NODAD>
        inet6 fe80::1%lo0/64 flags 0 scopeid 0x2
        inet 127.0.0.1/8 flags 0
```

The `virsh domiflist` command is also unable to locate the IP addresses altogether, so I get the feeling that I'm doing something semi-fishy here...

```
per@some-host:~/git/perlang-bsd-ci$ sudo virsh domiflist temp
 Interface   Type      Source           Model    MAC
--------------------------------------------------------------------
 vnet9       network   virt-lightning   virtio   52:54:00:84:9e:74

per@some-host:~/git/perlang-bsd-ci$ sudo virsh domiflist temp2
 Interface   Type      Source           Model    MAC
--------------------------------------------------------------------
 vnet10      network   virt-lightning   virtio   52:54:00:7a:b8:64

per@some-host:~/git/perlang-bsd-ci$ sudo virsh domifaddr temp --interface vnet9
 Name       MAC address          Protocol     Address
-------------------------------------------------------------------------------

per@some-host:~/git/perlang-bsd-ci$ sudo virsh domifaddr temp2 --interface vnet10
 Name       MAC address          Protocol     Address
-------------------------------------------------------------------------------
```

Any suggestions? I don't really understand how the machines can get the same IP address despite the MAC addresses being different. :thinking:

One way to workaround this would be to use static IP addresses, but if we can stick with DHCP for these ephemeral VMs I think it'll be much simpler. Once we have this resolved, I think I'll be able to adjust the shell scripts provided in the previously referred to [GitLab docs](https://docs.gitlab.com/runner/executors/custom_examples/libvirt.html) to successfully retrieve the IP address of the new VM.

Many thanks in advance!

**Comment by perlun** [comment]
Ah, these lines in `/etc/rc.conf` explain it I guess:

```ini
# You can safely remove the following lines starting with "cloud"
cloudinitlocal="YES"
cloudinit="YES"
cloudconfig="YES"
cloudfinal="YES"
hostname=netbsd-100-ufs
defaultroute=192.168.123.1
ifconfig_vioif0='192.168.123.5 netmask 255.255.255.0'
```

Removing them and rebooting the machine gives me no network at all though. :slightly_smiling_face: What's the proper way to re-query the DHCP server on the `virt-lightning` network on bootup?

**Comment by perlun** [comment]
I managed to get this working. The problem with "no network" was likely caused by incorrect usage of the `default` networking in `virt-clone`. At least, that's what I saw when later using `virt-install`; the `default` networking would pick up an existing bridge interface on the machine (which I created when I believed I would be using bridge networking for my VMs).. and there was no DHCP server available on that external network.

Anyway, I got this to the point where I can now create a NetBSD CI VM image which I use for my CI runs (https://gitlab.perlang.org/perlang/perlang/-/merge_requests/562). My scripts are based on `virt-lightning` + an Ansible role to install the dependencies into the role as needed; then, for each CI run I use the [libvirt](https://docs.gitlab.com/runner/executors/custom_examples/libvirt.html) scripts from GitLab with small modifications. All of this is available freely in my repo: https://gitlab.perlang.org/perlang/perlang-bsd-ci

I'll close this issue now, my problem has been solved. :tada:

**Comment by goneri** [comment]
@perlun thank you for the update and sorry for being slow. I'm glad to see you've managed to resolve your problem. And the tutorial is pretty cool and I'm super happy to see Virt-Lightning being part of your CI, Go Perland! :-).

**Comment by perlun** [comment]
Thank you @goneri, happy to be able to share some joy and nice to get some feedback from you, don't worry about the delay. :slightly_smiling_face: :raised_hands: Yeah, I just wanted "a way" to launch a BSD VM with reasonable ease, and stumbled upon your nice little https://bsd-cloud-image.org/ website. Great effort, thanks a lot for this. :bow: 

> I'm super happy to see Virt-Lightning being part of your CI, Go Perland! :-).

Thanks! :smiley: For those interested, here are some issues I created regarding BSD support for Perlang in general:

- FreeBSD: https://gitlab.perlang.org/perlang/perlang/-/issues/272
- NetBSD: https://gitlab.perlang.org/perlang/perlang/-/issues/502
- OpenBSD: https://gitlab.perlang.org/perlang/perlang/-/issues/503

I know it's a bit more work to upvote those because I'm a bit off-grid in using my own GitLab instance :grin: but I like the idea of "full autonomy" over my development. (I also like to use free software for developing free software. :heart: But I'll admit I use the non-free JetBrains IDEs since they are so good...)

**Comment by perlun** [comment]
(Btw, I presume "the tutorial" you mentioned was the README in https://gitlab.perlang.org/perlang/perlang-bsd-ci? It's not perfect, but it describes the steps needed to set this up. I might have missed some step though)

**Comment by perlun** [comment]
For reference, I checked my Firefox browser history and I might have found the website when googling for [`freebsd cloud image`](https://www.google.com/search?q=freebsd%20cloud%20image&ie=utf-8&oe=utf-8&client=firefox-b-m). One of my earlier thoughts was to set up some "short-lived" BSD VMs, perhaps recreating them once per day or so. I would then be running the `gitlab-runner` process inside the VM (and re-copying its `config.toml` to the VM each time it would be recreated). However, the architecture I ended up with was slightly different. I described it already above but here are the rough steps:

* Prepare an ephemeral VM using `virt-lightning`+Ansible using my repo mentioned above
* Modify the `libvirt` script [from the GitLab documentation](https://docs.gitlab.com/runner/executors/custom_examples/libvirt.html) to fit my use case
* Destroy the VM after each CI run.

(The main advantage of this is that each CI run is completely stateless. It's impossible for one CI run to affect another one, which provides great isolation between jobs.)

Despite all this overhead, a typical NetBSD CI run doesn't take more than 1m30s in my case: https://gitlab.perlang.org/perlang/perlang/-/jobs/906. It's probably not as fast as kicking up a Docker-based job, but it's still quite reasonable.
